src/Sarif.Sarifer/BackgroundAnalysisService.cs

```diff
@@ -17,37 +18,93 @@ namespace Microsoft.CodeAnalysis.Sarif.Sarifer
     [Export(typeof(IBackgroundAnalysisService))]
     internal class BackgroundAnalysisService : IBackgroundAnalysisService
     {
-#pragma warning disable CS0649 // Filled in by MEF
-#pragma warning disable IDE0044 // Assigned by MEF
+#pragma warning disable IDE0044, CS0649 // Provided by MEF
         [ImportMany]
         private IEnumerable<IBackgroundAnalyzer> analyzers;
-#pragma warning restore IDE0044
-#pragma warning restore CS0649
+
+        [ImportMany]
+        private IEnumerable<IBackgroundAnalysisSink> sinks;
+#pragma warning restore IDE0044, CS0649
 
         /// <inheritdoc/>
         public async Task AnalyzeAsync(string path, string text, CancellationToken cancellationToken)
         {
-            var tasks = new List<Task>(this.analyzers.Count());
+            var tasks = new List<Task<Stream>>(this.analyzers.Count());
 
             foreach (IBackgroundAnalyzer analyzer in this.analyzers)
             {
                 tasks.Add(analyzer.AnalyzeAsync(path, text, cancellationToken));
             }
 
-            await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false);
+            Stream[] streams = await Task.WhenAll(tasks).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
This needs to be done in a try/finally so we dispose the streams if there's an exception.

In fact there are two issues:

1. If an analyzer throws, analyzer base should dispose the stream.
2. If WriteStreamsToSinks throws, you need a try/finally here to ensure disposal.

I won't think this through in detail -- I leave that to you and @michaelcfanning -- but make sure you account for all of this.

Reply:
added a try/finally, always cleaning the streams.

---
In reply to: [535510544](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535510544) [](ancestors = 535510544)